source/opt/inline_pass.h

```diff
@@ -169,6 +166,52 @@ class InlinePass : public Pass {
   std::unordered_set<uint32_t> funcs_called_from_continue_;
 
  private:
+  // DebugInlinedAtContext keeps data used for generating DebugInlinedAt
+  // instructions.
+  class DebugInlinedAtContext {
+   public:
+    DebugInlinedAtContext(const Instruction* line, const DebugScope& scope)
+        : call_instr_line(line), call_instr_scope(scope){};
+
+    void UpdateInlinedAtContext(uint32_t callee_instr_inlined_at,
+                                uint32_t chain_head_id) {
+      callee_inlined_at2chain[callee_instr_inlined_at] = chain_head_id;
+    }
+
+    const Instruction* FunctionCallLine() { return call_instr_line; }
+    const DebugScope& FunctionCallScope() { return call_instr_scope; }
+    uint32_t InlinedAtChainHead(uint32_t callee_instr_inlined_at) {
+      auto chain_itr = callee_inlined_at2chain.find(callee_instr_inlined_at);
+      if (chain_itr != callee_inlined_at2chain.end()) return chain_itr->second;
+      return kNoInlinedAt;
+    }
+
+   private:
+    const Instruction*
+        call_instr_line;  // Line instruction of function call instruction
+    const DebugScope&
+        call_instr_scope;  // Debug scope of function call instruction
+
+    // Map from DebugInlinedAt ids of callee to head ids of new generated
+    // DebugInlinedAt chain.
+    std::unordered_map<uint32_t, uint32_t> callee_inlined_at2chain;
+  };
+
+  // Clone and map callee locals.  Return true if successful.
+  bool CloneAndMapLocals(Function* calleeFn,
+                         std::vector<std::unique_ptr<Instruction>>* new_vars,
+                         std::unordered_map<uint32_t, uint32_t>* callee2caller,
+                         DebugInlinedAtContext& inlined_at_ctx);
+
+  // Builds DebugScope for the callee instruction.
```

Comment:
You should be describing what is returned and any side effect.  You just say it builds something.  That tells me nothing about the return value or were the built debug scope will be.  You could try something like:

> Returns the debug scope corresponding to inlining |inlined_at_ctx| into the scope |callee_instr_scope|.  Generates all new debug instruction needed to represent the scope.